File: TotoRan/SingleChoiceViewController.m
//
//  SingleChoiceViewController.m
//  TotoRan
//
//  Created by 大山 孝 on 2015/05/04.
//  Copyright (c) 2015年 raksam.com. All rights reserved.
//

#import "SingleChoiceViewController.h"

#define WIDTH self.view.bounds.size.width

#define SIXPLUS_HEIGHT 700 //6+の画面高さを基準にする
#define BUTTON_HEIGHT 41.75 //ボタンの高さ
#define BUTTON_INTERVAL 5 //ボタンの間隔
#define TAG_INCREMENT 87 //tagの桁数増やし用
#define BASE_POINT 10 //起点
#define TAG_BASE 101 //tagの起点(101)
#define CLEAR_WIDTH 200 //選択解除ボタンwidth
#define CLEAR_HEIGHT 50 //選択解除ボタンheight

@interface SingleChoiceViewController ()

@end

@implementation SingleChoiceViewController
{
    SystemSoundID buttonSound; //ボタンの効果音
    SystemSoundID eraseButtonSound; //全消去ボタンの効果音
    AppDelegate *appDelegate; //情報取得用
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //画面タイトルを設定
    self.title = @"シングル選択";
    
    //ボタン効果音ファイル読み込み
    NSString *path = [[NSBundle mainBundle] pathForResource:@"kashan01" ofType:@"mp3"];
    NSURL *url = [NSURL fileURLWithPath:path];
    AudioServicesCreateSystemSoundID((CFURLRef)CFBridgingRetain(url), &buttonSound);
    
    //全消去ボタン効果音ファイル読み込み
    NSString *ePath = [[NSBundle mainBundle] pathForResource:@"erase" ofType:@"wav"];
    NSURL *eUrl = [NSURL fileURLWithPath:ePath];
    AudioServicesCreateSystemSoundID((CFURLRef)CFBridgingRetain(eUrl), &eraseButtonSound);
    
    //詳細設定画面への遷移ボタンを生成　navigationBarの右側ボタン
    UIBarButtonItem *pushButton = [[UIBarButtonItem alloc] initWithTitle:@"条件指定" style:UIBarButtonItemStyleBordered target:self action:@selector(push:)];
    pushButton.tag = 0;
    self.navigationItem.rightBarButtonItem = pushButton;
    
    //戻るボタンをカスタマイズ
    UIBarButtonItem *backButton = [[UIBarButtonItem alloc] initWithTitle:@"戻る" style:UIBarButtonItemStyleBordered target:self action:@selector(goBack)];
    self.navigationItem.leftBarButtonItem = backButton;
    
    //スクロールビューをself.viewにセット
    UIScrollView *scrollView = [UIScrollView new];
    scrollView.backgroundColor = [UIColor yellowColor];
    scrollView.frame = self.view.frame;
    scrollView.contentSize = CGSizeMake(WIDTH, SIXPLUS_HEIGHT); //6+の画面高さを基準にする
    scrollView.delegate = self;
    self.view = scrollView;
    
    //組み合わせ情報を取得
    appDelegate = [[UIApplication sharedApplication] delegate];
    NSArray *array = appDelegate.teamArray;
    
    //選択解除ボタンのy座標を格納する変数
    int clearY = 0;
    
    //ボタンのwidthを求める 「30」の内訳は、左右の余白１０×２　と　ボタン間隔５×２
    float buttonWidth = (WIDTH - 30) / 3.0;

    int tag = TAG_BASE; //tag
    int arrayCount = 0; //teamArrayの添え字用
    int xPoint = BASE_POINT; //起点x
    int yPoint = BASE_POINT; //起点y
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 13; j++) {
            if (i == 1) {
                [self makeButton:tag buttonRect:CGRectMake(xPoint, yPoint, buttonWidth, BUTTON_HEIGHT) teamName:@"DRAW"];
            } else {
                [self makeButton:tag buttonRect:CGRectMake(xPoint, yPoint, buttonWidth, BUTTON_HEIGHT) teamName:array[arrayCount]];
                arrayCount++;
            }
            yPoint = yPoint + BUTTON_HEIGHT + BUTTON_INTERVAL;
            tag++;
        }
        //選択解除ボタンのy座標を格納
        clearY = yPoint;
        
        //x起点を次の列にずらすのと、y起点のリセット、あとタグの桁を増やす(101,201,301)
        xPoint = xPoint + buttonWidth + BUTTON_INTERVAL;
        yPoint = BASE_POINT;
        tag += TAG_INCREMENT;
    }
    
    //選択解除ボタンを配置
    UIButton *clearButton = [[UIButton alloc] initWithFrame:CGRectMake(WIDTH / 2 - 100, clearY + 20, CLEAR_WIDTH, CLEAR_HEIGHT)];
    [clearButton setTitle:@"選択解除" forState:UIControlStateNormal];
    [clearButton setTitleColor:[UIColor blueColor] forState:UIControlStateNormal];
    [clearButton setBackgroundImage:[UIImage imageNamed:@"nomal.png"] forState:UIControlStateNormal];
    [clearButton setBackgroundImage:[UIImage imageNamed:@"nomal.png" ] forState:UIControlStateSelected];
    [clearButton addTarget:self action:@selector(clear:) forControlEvents:UIControlEventTouchUpInside];
    [clearButton.titleLabel setFont:[UIFont systemFontOfSize:appDelegate.fontSize + 2]];
    //縁取り
    [[clearButton layer] setBorderColor:[[UIColor blueColor] CGColor]]; // 枠線の色
    [[clearButton layer] setBorderWidth:1.0]; // 枠線の太さ
    [self.view addSubview:clearButton];

}


//ボタン生成メソッド
- (void)makeButton:(int)tag buttonRect:(CGRect)buttonRect teamName:(NSString *)teamName
{
    UIButton *button = [[UIButton alloc] initWithFrame:buttonRect];
    [button setTitle:teamName forState:UIControlStateNormal];
    [button setTitleColor:[UIColor blueColor] forState:UIControlStateNormal];
    [button setBackgroundImage:[UIImage imageNamed:@"nomal.png"] forState:UIControlStateNormal];
    [button setBackgroundImage:[UIImage imageNamed:@"select_single.png" ] forState:UIControlStateSelected];
    button.tag = tag;
    [button addTarget:self action:@selector(toggle:) forControlEvents:UIControlEventTouchDown];
    [button.titleLabel setFont:[UIFont systemFontOfSize:appDelegate.fontSize + 2]];
    //縁取り
    [[button layer] setBorderColor:[[UIColor blueColor] CGColor]]; // 枠線の色
    [[button layer] setBorderWidth:1.0]; // 枠線の太さ
    [self.view addSubview:button];
}

//ラジオボタン
- (void)toggle:(UIButton *)button
{
    //自分自身はトグル
    button.selected = !button.selected;
    
    //トグルに合わせて文字色を変更(青か白)
    if (button.selected) {
        [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    } else {
        [button setTitleColor:[UIColor blueColor] forState:UIControlStateNormal];
    }
    
    //自分の仲間の選択があれば解除するために下二桁を取得(下二桁が同一)
    NSString *tagSuffixString = [[NSString stringWithFormat:@"%ld", (long)button.tag] substringFromIndex:1];
    int tagNum = tagSuffixString.intValue + 100;
    
    //お仲間を探すために３回ループ
    for (int i = tagNum; i < 314; i = i + 100) {
        //自分自身じゃ無いボタンを非選択状態にする
        if (i != button.tag) {
            UIButton *targetButton = (UIButton *)[self.view viewWithTag:i];
            targetButton.selected = NO;
            [targetButton setTitleColor:[UIColor blueColor] forState:UIControlStateNormal];
        }
    }
    
    //効果音
    AudioServicesPlaySystemSound(buttonSound);
}

//クリアボタン
- (void)clear:(UIButton *)button
{
    int tag = TAG_BASE; //tag
    for (int i = 0; i < 13; i++) {
        for (int j = 0; j < 3; j++) {
            UIButton *targetButton = (UIButton *)[self.view viewWithTag:tag];
            targetButton.selected = NO;
            [targetButton setTitleColor:[UIColor blueColor] forState:UIControlStateNormal];
            tag = tag + 100;
        }
        tag = tag - 299;
    }
    
    //効果音
    AudioServicesPlaySystemSound(eraseButtonSound);
}

//次画面への遷移
- (void)push:(UIButton *)button
{
    //次の画面へボタンの選択状態を伝えるためにUiButtonを格納した２次元Arrayを作成
    NSMutableArray *arrayParent = [NSMutableArray array];
    
    int tag = TAG_BASE; //tag
    int choiceCountCheck = 0; //全て選択されてたらランダム不要だよね
    for (int i = 0; i < 13; i++) {
        NSMutableArray *arrayChiled = [NSMutableArray array];
        for (int j = 0; j < 3; j++) {
            UIButton *targetButton = (UIButton *)[self.view viewWithTag:tag];
            [arrayChiled addObject:targetButton];
            tag = tag + 100;
            if (targetButton.selected) choiceCountCheck++;
        }
        [arrayParent addObject:arrayChiled];
        tag = tag - 299;
    }

    //全て選択されてたらランダム不要だよね
    if (choiceCountCheck == 13) {
        [[[UIAlertView alloc] initWithTitle:@""
                                    message:@"もうその組み合わせで買えば？"
                                   delegate:self
                          cancelButtonTitle:@"OK"
                          otherButtonTitles:nil] show];
        //以降のスキップ
        return;
    }
    
    
    SingleDetailsViewController *sdvc = [SingleDetailsViewController new];
    // Tab bar を非表示
    sdvc.hidesBottomBarWhenPushed = YES;
    
    //次の画面へUiButtonの二次元Arrayを受け渡し
    sdvc.buttonArray = arrayParent;

    [self.navigationController pushViewController:sdvc animated:YES];
}

//シングル or マルチ選択画面に戻る
- (void)goBack
{
    //選択画面で何も選択されてなかったら警告無しで戻しちゃっていいんじゃね？
    
    
    
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"確認"
                                message:@"前画面に戻ると現在の選択状態が破棄されますが、よろしいですか？"
                               delegate:self
                      cancelButtonTitle:@"いいえ"
                      otherButtonTitles:@"はい", nil];
    alert.delegate = self;
    [alert show];
}

//はいが押されたらシングル or マルチ選択画面に戻る
-(void) alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex
{
    if(buttonIndex != alertView.cancelButtonIndex) [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)didReceiveMemoryWarning {[super didReceiveMemoryWarning];}


@end

File: TotoRan/SingleChoiceViewController.swift
//
//  SingleChoiceViewController.swift
//  TotoRan
//
//  Created by 大山 孝 on 2015/12/07.
//  Copyright © 2015年 raksam.com. All rights reserved.
//

import UIKit

class SingleChoiceViewController: ChoiceViewController
{
    //init
    override func setup() {
        //ボタン類のベースカラーを指定
        baseColor = UIColor.blueColor()
        
        //pngファイルを指定
        pngFile = UIImage(named: FILE_NAME.SINCLE_PNG)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //画面タイトルを設定
        self.title = "シングル選択"
    }
    
    //ボタン選択時の挙動　シングルのみ同一枠内で１つしか選択できないロジック追加
    override func toggle(button: UIButton) {
        //同一部分の処理は親クラスのメソッド機能を呼び出す
        super.toggle(button)
        
        //同一枠を識別するためのtag下２桁を抽出
        let tagSuffixString = (button.tag.description as NSString).substringFromIndex(1)
        let tagNum = Int(tagSuffixString)! + CHOICE.TAGMAXBASE
        
        //同一枠をループ指定
        for var i = tagNum; i < CHOICE.TAGMAX; i = i + CHOICE.TAGMAXBASE {
            //自分以外の同一枠を非選択状態にする
            if i != button.tag {
                let target = view.viewWithTag(i) as! UIButton
                target.selected = false
                target.setTitleColor(baseColor, forState: .Normal)
            }
        }
    }
    
    //次画面への遷移
    override func push() {
        //次画面へボタンの選択状態を渡す1次元Arrayを作成
        var arrayParent: [Int] = []
        var tag = CHOICE.TAG_BASE
        
        //１枠内で選択されてた場合はその値(1,0,2)を、選択されてなかった場合は「9」をセットする
        for var i = 0; i < 13; i++ {
            //１枠内の３つの数字を一時的に格納する
            var threeValArray: [Int] = []
            for var j = 0; j < 3; j++ {
                //ボタンを取得
                let target = view.viewWithTag(tag) as! UIButton
                var setTag = 9
                if target.selected {
                    let tagStr = String(tag)
                    setTag = Int(String(tagStr[tagStr.startIndex]))!
                    if setTag == 2 {
                        setTag = 0
                    } else if setTag == 3 {
                        setTag = 2
                    }
                }
                threeValArray.append(setTag)
                tag += 100
            }
            //9しかなければ9を、それ以外があればその値をarrayParentに格納
            threeValArray = NSOrderedSet(array: threeValArray).array as! [Int] //まずは重複削除
            arrayParent.append(threeValArray.sort(<).first!) //昇順ソートして最初の要素を格納
            tag -= 299
        }
        
        //非選択枠の数を取得
        let nonTapCount = arrayParent.filter({ (val: Int) -> Bool in return val == 9}).count
        
        
        //全て選択されてたらランダム処理は不要　ジョークアプリって言われちまったよ！！！！！www
//        if nonTapCount == 0 {
//            alertDisplay(MESSAGE_STR.ALL_SELECTED)
//            return
//        }
        
        //各選択枠の数を取得
        let homeCount = arrayParent.filter({ (val: Int) -> Bool in return val == 1}).count
        let drawCount = arrayParent.filter({ (val: Int) -> Bool in return val == 0}).count
        let awayCount = arrayParent.filter({ (val: Int) -> Bool in return val == 2}).count
        
        //口数のmax値を算出
        var unitMax = 10
        if nonTapCount == 0 {
            unitMax = 1
        } else if nonTapCount == 1 {
            unitMax = 3
        } else if nonTapCount == 2 {
            unitMax = 9
        }
        
        //詳細画面への遷移
        let sdvc = SingleDetailsViewController()
        sdvc.hidesBottomBarWhenPushed = true
        
        //13のInt(1 or 0 or 2 or 9)の一次元配列
        sdvc.arrayParent = arrayParent
        
        //ダブルとトリプルそれぞれのピッカーに表示する数列文字列の２次元配列
        sdvc.pickerItemArray = [
            itemSet(homeCount, max: arrayParent.count - (drawCount + awayCount)),
            itemSet(drawCount, max: arrayParent.count - (homeCount + awayCount)),
            itemSet(awayCount, max: arrayParent.count - (homeCount + drawCount)),
            itemSet(1, max: unitMax)
        ]

        navigationController?.pushViewController(sdvc, animated: true)
    }
}
